### PR TITLE
Support middlewares for subscriptions

### LINKED ISSUE #188: Can subscriptions have middleware?
In `execute` you can pass a `MiddlewareManager`  is there an alternative for `subscribe`?
I want to run code before a resolver is called.

### PR DESCRIPTION
**Pull request by nrbnlulu**
Fix #188

I am not sure if this requires more tests because AFAIR its basically the same as async execution.
Please let me know it needs more @Cito

### CODE REVIEW

**Comment by nrbnlulu** [comment]
@Cito can you aprove the workflows?

**Comment by Cito** [comment]
Thanks, will review this tomorrow.

File: tests/execution/test_middleware.py [hunk 1]
```diff
@@ -236,6 +245,39 @@ async def resolve(self, next_, *args, **kwargs):
...
+
+            async def bar_resolve(_obj, _info):
+                yield "bar"
+                yield "oof"
+            test_type = GraphQLObjectType(
+                "Subscription",
+                {
+                    "bar": GraphQLField(
+                        GraphQLString,
+                        resolve=lambda message, _: message,
```
Comment (Cito) [inline tests/execution/test_middleware.py]:
```suggestion
                        resolve=lambda message, _info: message,
```

File: tests/execution/test_middleware.py [hunk 2]
```diff
@@ -236,6 +245,39 @@ async def resolve(self, next_, *args, **kwargs):
             result = await awaitable_result
             assert result.data == {"field": "devloseR"}
 
+        @pytest.mark.asyncio()
+        async def subscription_simple():
+
+            async def bar_resolve(_obj, _info):
+                yield "bar"
+                yield "oof"
+            test_type = GraphQLObjectType(
```
Comment (Cito) [inline tests/execution/test_middleware.py:246]:
```suggestion

            test_type = GraphQLObjectType(
```

File: tests/execution/test_middleware.py [hunk 3]
```diff
@@ -1,11 +1,20 @@
+import inspect
 from typing import Awaitable, cast
 
 import pytest
+from graphql import subscribe
```
Comment (Cito) [inline tests/execution/test_middleware.py]:
```suggestion
from graphql.execution import Middleware, MiddlewareManager, execute, subscribe
```

File: tests/execution/test_middleware.py [hunk 4]
```diff
@@ -1,11 +1,20 @@
...
 from graphql.execution import Middleware, MiddlewareManager, execute
```
Comment (Cito) [inline tests/execution/test_middleware.py]:
```suggestion
```

File: tests/execution/test_middleware.py [hunk 5]
```diff
@@ -236,6 +245,39 @@ async def resolve(self, next_, *args, **kwargs):
...
+            agen = subscribe(
+                _create_subscription_schema(test_type),
+                doc,
+                middleware=MiddlewareManager(reverse_middleware),
+            )
+            assert inspect.isasyncgen(agen)
+            data = (await agen.__anext__()).data
+            assert data == {"bar": "rab"}
+            data = (await agen.__anext__()).data
+            assert data == {"bar": "foo"}
```
Comment (Cito) [inline tests/execution/test_middleware.py]:
```suggestion

            async def reverse_middleware(next_, *args, **kwargs):
                return next_(*args, **kwargs)[::-1]

            noop_type = GraphQLObjectType(
                "Noop",
                {"noop": GraphQLField(GraphQLString, resolve=lambda *_args: "noop")},
            )
            schema = GraphQLSchema(query=noop_type, subscription=test_type)

            subscription = subscribe(
                schema,
                parse("subscription { bar }"),
                middleware=MiddlewareManager(reverse_middleware),
            )
            assert inspect.isasyncgen(subscription)
            data = (await subscription.__anext__()).data
            assert data == {"bar": "rab"}
            data = (await subscription.__anext__()).data
            assert data == {"bar": "foo"}
```

File: tests/execution/test_middleware.py [hunk 6]
```diff
@@ -1,11 +1,20 @@
...
 from typing import Awaitable, cast
 
 import pytest
+from graphql import subscribe
 from graphql.execution import Middleware, MiddlewareManager, execute
 from graphql.language.parser import parse
 from graphql.type import GraphQLField, GraphQLObjectType, GraphQLSchema, GraphQLString
 
 
+def _create_subscription_schema(tp: GraphQLObjectType) -> GraphQLSchema:
```
Comment (Cito) [inline tests/execution/test_middleware.py]: Since this is only used in one test, better move it there (see suggestion below).

**Review by Cito — CHANGES_REQUESTED** [review]
Looks good to me, thank you.

Just minor stylistic changes for consistency, see suggestions.

File: tests/execution/test_middleware.py [hunk 7]
```diff
@@ -236,6 +237,45 @@ async def resolve(self, next_, *args, **kwargs):
...
+            )
+            doc = parse("subscription { bar }")
+
+            async def reverse_middleware(next_, value, info, **kwargs):
+                awaitable_maybe = next_(value, info, **kwargs)
+                return awaitable_maybe[::-1]
+
+            noop_type = GraphQLObjectType(
+                "Noop",
+                {"noop": GraphQLField(GraphQLString, resolve=lambda *_args: "noop")},
```
Comment (Cito) [inline tests/execution/test_middleware.py]:
```suggestion
                {"noop": GraphQLField(GraphQLString)},
```

Comment (Cito) [reply tests/execution/test_middleware.py]: This is not needed and should also solve the coverage issue.

**Review by Cito — CHANGES_REQUESTED** [review]
The dummy resolver can be removed.

**Review by Cito — APPROVED** [review]

**Pull request merged**